File: frontend/src/services/submissions.js
// =============================================================================
// Submissions Service — frontend/src/services/submissions.js
// =============================================================================
// API calls for the material submission feature.
//
// Public:
//   createSubmission(formData)  — POST /api/v1/submissions  (multipart)
//
// Admin (requires Firebase token in Authorization header):
//   getSubmissions(token, status)               — GET  /api/v1/submissions
//   getSubmission(token, id)                    — GET  /api/v1/submissions/{id}
//   approveSubmission(token, id, body)           — POST /api/v1/submissions/{id}/approve
//   rejectSubmission(token, id, body)            — POST /api/v1/submissions/{id}/reject
//   restoreSubmission(token, id)                 — POST /api/v1/submissions/{id}/restore
//   downloadSubmissionFile(token, fileId)        — GET  /api/v1/submissions/files/{fileId}/download
// =============================================================================

import { apiFetch } from '../lib/api'
import { getFirebaseToken } from '../lib/firebase'

function getApiUrl(path) {
  const envUrl = import.meta.env.VITE_API_BASE_URL || ''
  if (envUrl.includes('onrender.com') || envUrl.includes('render.com')) {
    return path
  }
  if (typeof window !== 'undefined' && (window.location.hostname.endsWith('vercel.app') || window.location.hostname === 'tn-board-portal.vercel.app')) {
    return path
  }
  return `${envUrl}${path}`
}

/**
 * Submit educational material (requires auth).
 *
 * @param {FormData} formData  Must include: publisher_name, files[].
 *                             Optional: details.
 * @returns {Promise<{ id: string, status: string, message: string }>}
 */
export async function createSubmission(formData) {
  const token = await getFirebaseToken()
  if (!token) throw new Error('Authentication required to submit material')

  // We use raw fetch here (not apiFetch) because we need multipart/form-data
  // without setting Content-Type manually — the browser sets boundary automatically.
  const url = getApiUrl('/api/v1/submissions')
  const response = await fetch(url, {
    method: 'POST',
    body: formData,
    headers: {
      'Authorization': `Bearer ${token}`
    }
    // DO NOT set Content-Type — browser sets it with correct boundary
  })

  if (!response.ok) {
    let detail = `HTTP ${response.status}`
    try {
      const body = await response.json()
      detail = body?.detail || body?.message || detail
    } catch {
      // ignore JSON parse errors
    }
    throw new Error(detail)
  }

  return response.json()
}

// ── Contributor: Get authenticated user's own submissions ────────────────────


/**
 * Fetch all submissions submitted by the current authenticated user.
 * @returns {Promise<{ data: UserSubmissionItem[], total: number }>}
 */
export async function getMySubmissions() {
  const token = await getFirebaseToken()
  if (!token) throw new Error('Authentication required')
  return apiFetch('/api/v1/submissions/my', {
    headers: { Authorization: `Bearer ${token}` },
  })
}

// ── Admin: List submissions ───────────────────────────────────────────────────

/**
 * List all submissions (admin only).
 *
 * Supports both getSubmissions(statusFilter, limit) and legacy getSubmissions(token, statusFilter, limit).
 *
 * @param {string|null} [tokenOrStatusFilter]  statusFilter ('pending' | 'approved' | 'rejected' | null) or legacy token
 * @param {string|number|null} [statusFilterOrLimit]
 * @param {number} [maybeLimit]  Max results (default 50)
 * @returns {Promise<{ data: SubmissionListItem[], count: number, status_filter: string|null }>}
 */
export async function getSubmissions(tokenOrStatusFilter = null, statusFilterOrLimit = null, maybeLimit = 50) {
  let statusFilter = null
  let limit = 50

  if (typeof tokenOrStatusFilter === 'string' && (tokenOrStatusFilter.startsWith('ey') || tokenOrStatusFilter.length > 50)) {
    statusFilter = statusFilterOrLimit
    limit = typeof maybeLimit === 'number' ? maybeLimit : 50
  } else {
    statusFilter = tokenOrStatusFilter
    limit = typeof statusFilterOrLimit === 'number' ? statusFilterOrLimit : 50
  }

  const token = await getFirebaseToken()
  if (!token) throw new Error('Authentication required')

  const params = new URLSearchParams({ limit: String(limit) })
  if (statusFilter) params.set('status', statusFilter)
  return apiFetch(`/api/v1/submissions?${params}`, {
    headers: { Authorization: `Bearer ${token}` },
  })
}

// ── Admin: Get submission detail ──────────────────────────────────────────────

/**
 * Get a single submission with its file list (admin only).
 *
 * Supports both getSubmission(id) and legacy getSubmission(token, id).
 *
 * @param {string} tokenOrId  Submission UUID or legacy token
 * @param {string} [maybeId]  Submission UUID if token was passed first
 * @returns {Promise<SubmissionOut>}
 */
export async function getSubmission(tokenOrId, maybeId) {
  const id = maybeId !== undefined ? maybeId : tokenOrId
  const token = await getFirebaseToken()
  if (!token) throw new Error('Authentication required')

  return apiFetch(`/api/v1/submissions/${id}`, {
    headers: { Authorization: `Bearer ${token}` },
  })
}

// ── Admin: Approve submission ─────────────────────────────────────────────────

/**
 * Approve a pending submission and create paper record(s).
 * Obtains a fresh Firebase ID token immediately before sending the mutation request.
 *
 * Supports both approveSubmission(id, body) and legacy approveSubmission(token, id, body).
 *
 * @param {string} tokenOrId  Submission UUID or legacy token
 * @param {string|object} idOrBody Submission UUID or body object
 * @param {object} [maybeBody] Body if token was passed as first argument
 * @returns {Promise<{ submission_id: string, status: string, paper_ids: number[] }>}
 */
export async function approveSubmission(tokenOrId, idOrBody, maybeBody) {
  let id, body
  if (maybeBody !== undefined) {
    id = idOrBody
    body = maybeBody
  } else {
    id = tokenOrId
    body = idOrBody
  }

  const token = await getFirebaseToken(true)
  if (!token) throw new Error('Authentication required')

  return apiFetch(`/api/v1/submissions/${id}/approve`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  })
}

// ── Admin: Reject submission ──────────────────────────────────────────────────

/**
 * Reject a pending submission.
 * Obtains a fresh Firebase ID token immediately before sending the mutation request.
 *
 * Supports both rejectSubmission(id, body) and legacy rejectSubmission(token, id, body).
 *
 * @param {string} tokenOrId  Submission UUID or legacy token
 * @param {string|object} idOrBody Submission UUID or body object
 * @param {object} [maybeBody] Body if token was passed as first argument
 * @returns {Promise<{ submission_id: string, status: string, rejection_reason: string|null }>}
 */
export async function rejectSubmission(tokenOrId, idOrBody, maybeBody) {
  let id, body
  if (maybeBody !== undefined) {
    id = idOrBody
    body = maybeBody
  } else {
    id = tokenOrId
    body = idOrBody
  }

  const token = await getFirebaseToken(true)
  if (!token) throw new Error('Authentication required')

  return apiFetch(`/api/v1/submissions/${id}/reject`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  })
}

// ── Admin: Restore rejected submission to pending ─────────────────────────────

/**
 * Restore a rejected submission back to pending.
 * Obtains a fresh Firebase ID token immediately before sending the mutation request.
 *
 * Supports both restoreSubmission(id) and legacy restoreSubmission(token, id).
 *
 * @param {string} tokenOrId  Submission UUID or legacy token
 * @param {string} [maybeId]  Submission UUID if token was passed first
 * @returns {Promise<{ submission_id: string, status: string }>}
 */
export async function restoreSubmission(tokenOrId, maybeId) {
  const id = maybeId !== undefined ? maybeId : tokenOrId
  const token = await getFirebaseToken(true)
  if (!token) throw new Error('Authentication required')

  return apiFetch(`/api/v1/submissions/${id}/restore`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({}),
  })
}

// ── Admin: Download a private submission file via backend proxy ───────────────

/**
 * Download a private submission file through the backend proxy endpoint.
 *
 * Supports both downloadSubmissionFile(fileId, filename) and legacy downloadSubmissionFile(token, fileId, filename).
 *
 * @param {string} tokenOrFileId  File UUID or legacy token
 * @param {string} fileIdOrFilename Original filename or file UUID
 * @param {string} [maybeFilename] Original filename if token was passed first
 * @returns {Promise<void>}  Triggers browser file save
 */
export async function downloadSubmissionFile(tokenOrFileId, fileIdOrFilename, maybeFilename) {
  let fileId, filename
  if (maybeFilename !== undefined) {
    fileId = fileIdOrFilename
    filename = maybeFilename
  } else {
    fileId = tokenOrFileId
    filename = fileIdOrFilename
  }

  const token = await getFirebaseToken()
  if (!token) throw new Error('Authentication required')

  const url = getApiUrl(`/api/v1/submissions/files/${fileId}/download`)
  const response = await fetch(url, {
    method: 'GET',
    headers: { Authorization: `Bearer ${token}` },
  })

  if (!response.ok) {
    let detail = `HTTP ${response.status}`
    try {
      const resBody = await response.json()
      detail = resBody?.detail || resBody?.message || detail
    } catch {
      // ignore — response may not be JSON for binary errors
    }
    throw new Error(detail)
  }

  const blob = await response.blob()
  const objectUrl = URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = objectUrl
  link.download = filename || 'download'
  document.body.appendChild(link)
  link.click()
  document.body.removeChild(link)
  setTimeout(() => URL.revokeObjectURL(objectUrl), 10000)
}

File: backend/app/api/v1/endpoints/submissions.py
"""
Submissions endpoints — POST /api/v1/submissions (public)
                     — GET  /api/v1/submissions (admin)
                     — GET  /api/v1/submissions/{id} (admin)
                     — GET  /api/v1/submissions/files/{id}/download (admin)
                     — POST /api/v1/submissions/{id}/approve (admin)
                     — POST /api/v1/submissions/{id}/reject (admin)
                     — POST /api/v1/submissions/{id}/restore (admin)

Route layer responsibilities:
  - Validate parameters (FastAPI type hints)
  - Verify auth (Depends(require_admin))
  - Call SubmissionsService
  - Return response
  - Nothing else

All business logic lives in SubmissionsService.
All database access lives in SubmissionsRepository.
"""

import logging
import urllib.parse
from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile, status
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.dependencies.auth import get_current_user, require_admin, require_role
from app.dependencies.supabase import get_db
from app.schemas.submission import (
    ApproveRequest,
    RejectRequest,
    SubmissionCreateResponse,
    SubmissionDeleteResponse,
    SubmissionListResponse,
    SubmissionOut,
    UserSubmissionsResponse,
)
from app.services.submissions_service import SubmissionsService
from app.utils.exceptions import DatabaseError, NotFoundError, ValidationError

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/submissions", tags=["Submissions"])


# ── POST /api/v1/submissions — authenticated (contributor or higher) ──────────


@router.post(
    "",
    response_model=SubmissionCreateResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Submit educational material",
    description=(
        "Authenticated endpoint. Accepts a multipart form with contributor details and "
        "up to 5 files (PDF, Word, or image). "
        "The submission is stored with status='pending' for admin review.\n\n"
        "**Auth required.** Submissions do NOT immediately appear on the public site."
    ),
    responses={
        201: {"description": "Submission created, pending admin review"},
        422: {"description": "Validation error (missing fields, bad file type, file too large, etc.)"},
    },
)
async def create_submission(
    publisher_name: Annotated[str, Form(description="Contributor or publisher name")],
    files: Annotated[
        list[UploadFile],
        File(description="One or more files (PDF, DOC, DOCX, JPG, PNG). Max 5 files, 25 MB each."),
    ],
    details: Annotated[
        str | None,
        Form(description="Optional description or additional details"),
    ] = None,
    current_user: dict = Depends(require_role(["USER", "CONTRIBUTOR", "ADMIN", "SUPER_ADMIN"])),
    db: Session = Depends(get_db),
) -> SubmissionCreateResponse:
    """
    Create a new material submission.
    """
    service = SubmissionsService(db)
    return await service.create_submission(
        publisher_name=publisher_name,
        email=current_user.get("email"),
        firebase_uid=current_user.get("firebase_uid"),
        details=details,
        files=files,
    )


# ── GET /api/v1/submissions/my — authenticated user's own submissions ──────────


@router.get(
    "/my",
    response_model=UserSubmissionsResponse,
    status_code=status.HTTP_200_OK,
    summary="Get current user's submitted materials",
    description=(
        "Authenticated endpoint. Returns all submissions uploaded by the current user, "
        "including attached file names, current status (Under Review, Published, Rejected), "
        "rejection reason if rejected, and links to published papers if approved."
    ),
    responses={
        200: {"description": "User's submissions"},
        401: {"description": "Authentication required"},
    },
)
async def get_my_submissions(
    current_user: dict = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> UserSubmissionsResponse:
    """Return all submissions made by the authenticated user."""
    service = SubmissionsService(db)
    return service.get_user_submissions(
        firebase_uid=current_user["firebase_uid"],
        email=current_user.get("email"),
    )


# ── GET /api/v1/submissions — admin ──────────────────────────────────────────



@router.get(
    "",
    response_model=SubmissionListResponse,
    summary="List material submissions (admin)",
    description=(
        "Admin only. Returns submissions ordered by submitted date (newest first).\n\n"
        "Use the `status` query param to filter: `pending`, `approved`, or `rejected`."
    ),
    responses={
        200: {"description": "List of submissions"},
        401: {"description": "No auth token provided"},
        403: {"description": "Invalid or expired token"},
    },
)
async def list_submissions(
    status_filter: Annotated[
        str | None,
        Query(alias="status", description="Filter by status: pending | approved | rejected"),
    ] = None,
    limit: Annotated[
        int,
        Query(description="Max submissions to return", ge=1, le=200),
    ] = 50,
    current_user: dict = Depends(require_admin),
    db: Session = Depends(get_db),
) -> SubmissionListResponse:
    """Return all submissions for the admin review UI."""
    service = SubmissionsService(db)
    return service.list_submissions(status=status_filter, limit=limit)


# ── GET /api/v1/submissions/files/{file_id}/download — admin ─────────────────
#
# NOTE: This route MUST be declared BEFORE /{submission_id} so that FastAPI
# matches /submissions/files/… before it tries to capture 'files' as a
# {submission_id} path parameter.


@router.get(
    "/files/{file_id}/download",
    summary="Download a private submission file (admin)",
    description=(
        "Admin only. Proxies the private Supabase Storage file through the backend "
        "so the browser receives a proper Content-Disposition: attachment response.\n\n"
        "The HTML `download` attribute is silently ignored by browsers for cross-origin "
        "URLs, which is why signed URLs cannot be used for downloads directly. "
        "This endpoint solves that by streaming the file through the backend.\n\n"
        "The service-role key is never exposed to the browser."
    ),
    responses={
        200: {"description": "File bytes with Content-Disposition: attachment"},
        401: {"description": "No auth token provided"},
        403: {"description": "Invalid or expired token"},
        404: {"description": "File not found"},
        500: {"description": "Storage retrieval failed"},
    },
)
async def download_submission_file(
    file_id: str,
    current_user: dict = Depends(require_admin),
    db: Session = Depends(get_db),
) -> Response:
    """
    Proxy-download a submission file from the private Supabase bucket.

    Returns the raw bytes with:
      Content-Type:        <appropriate MIME type>
      Content-Disposition: attachment; filename="<original_filename>"

    The browser will save the file locally regardless of origin.
    """
    service = SubmissionsService(db)
    try:
        file_bytes, content_type, original_filename = service.download_file(file_id)
    except NotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Submission file '{file_id}' not found.",
        )
    except (DatabaseError, RuntimeError) as exc:
        logger.error("Download proxy failed for file_id=%s: %s", file_id, exc)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve the file from storage.",
        )

    # RFC 5987 encoding for non-ASCII filenames
    encoded_name = urllib.parse.quote(original_filename, safe="")
    disposition = f'attachment; filename="{original_filename}"; filename*=UTF-8\'\'{encoded_name}'

    return Response(
        content=file_bytes,
        media_type=content_type,
        headers={"Content-Disposition": disposition},
    )


# ── GET /api/v1/submissions/{id} — admin ─────────────────────────────────────


@router.get(
    "/{submission_id}",
    response_model=SubmissionOut,
    summary="Get submission detail (admin)",
    description=(
        "Admin only. Returns one submission with its full file list.\n\n"
        "Use this to inspect files before approving or rejecting."
    ),
    responses={
        200: {"description": "Submission detail with files"},
        401: {"description": "No auth token provided"},
        403: {"description": "Invalid or expired token"},
        404: {"description": "Submission not found"},
    },
)
async def get_submission(
    submission_id: str,
    current_user: dict = Depends(require_admin),
    db: Session = Depends(get_db),
) -> SubmissionOut:
    """Return one submission with its file list."""
    service = SubmissionsService(db)
    return service.get_submission(submission_id)


# ── POST /api/v1/submissions/{id}/approve — admin ────────────────────────────


@router.post(
    "/{submission_id}/approve",
    status_code=status.HTTP_200_OK,
    summary="Approve a submission (admin)",
    description=(
        "Admin only. Approves a pending submission and creates one published "
        "paper record in the `papers` table for each submitted file.\n\n"
        "The request body must supply `subject_id`, `exam_type`, `year`, and "
        "`paper_type` — the admin fills these in at approval time because the "
        "public form does not collect subject/class information.\n\n"
        "The submission status is set to `approved` and `reviewed_at` is set to now."
    ),
    responses={
        200: {"description": "Approved — paper(s) created"},
        401: {"description": "No auth token"},
        403: {"description": "Invalid token"},
        404: {"description": "Submission not found"},
        422: {"description": "Submission is not pending, or validation error"},
    },
)
async def approve_submission(
    submission_id: str,
    req: ApproveRequest,
    current_user: dict = Depends(require_admin),
    db: Session = Depends(get_db),
) -> dict:
    """Approve a pending submission and create paper records."""
    service = SubmissionsService(db)
    return service.approve_submission(submission_id, req)


# ── POST /api/v1/submissions/{id}/reject — admin ─────────────────────────────


@router.post(
    "/{submission_id}/reject",
    status_code=status.HTTP_200_OK,
    summary="Reject a submission (admin)",
    description=(
        "Admin only. Rejects a pending submission.\n\n"
        "No paper record is created. The uploaded files remain in storage. "
        "An optional `rejection_reason` can be stored for admin reference.\n\n"
        "The submission status is set to `rejected` and `reviewed_at` is set to now.\n\n"
        "A rejected submission can later be restored to `pending` via the "
        "`POST /submissions/{id}/restore` endpoint."
    ),
    responses={
        200: {"description": "Rejected — no paper created"},
        401: {"description": "No auth token"},
        403: {"description": "Invalid token"},
        404: {"description": "Submission not found"},
        422: {"description": "Submission is not pending"},
    },
)
async def reject_submission(
    submission_id: str,
    req: RejectRequest,
    current_user: dict = Depends(require_admin),
    db: Session = Depends(get_db),
) -> dict:
    """Reject a pending submission."""
    service = SubmissionsService(db)
    return service.reject_submission(submission_id, req)


# ── POST /api/v1/submissions/{id}/restore — admin ────────────────────────────


@router.post(
    "/{submission_id}/restore",
    status_code=status.HTTP_200_OK,
    summary="Restore a rejected submission to pending (admin)",
    description=(
        "Admin only. Moves a `rejected` submission back to `pending` so it can "
        "be reviewed and approved.\n\n"
        "Clears `rejection_reason` and `reviewed_at`.\n\n"
        "Only `rejected` submissions can be restored. "
        "`approved` submissions cannot be restored (they are already published).\n\n"
        "State machine: REJECTED → PENDING → APPROVED"
    ),
    responses={
        200: {"description": "Submission restored to pending"},
        401: {"description": "No auth token"},
        403: {"description": "Invalid token"},
        404: {"description": "Submission not found"},
        422: {"description": "Submission is not rejected"},
    },
)
async def restore_submission(
    submission_id: str,
    current_user: dict = Depends(require_admin),
    db: Session = Depends(get_db),
) -> dict:
    """Restore a rejected submission back to pending for re-review."""
    service = SubmissionsService(db)
    return service.restore_submission(submission_id)


# ── DELETE /api/v1/submissions/{id} — admin ──────────────────────────────────


@router.delete(
    "/{submission_id}",
    response_model=SubmissionDeleteResponse,
    status_code=status.HTTP_200_OK,
    summary="Delete a submission (admin)",
    description=(
        "Admin only. Permanently deletes a pending or rejected submission, "
        "its submission_files records, and its uploaded private files in Supabase Storage.\n\n"
        "Approved submissions linked to published papers cannot be deleted."
    ),
    responses={
        200: {"description": "Submission deleted successfully"},
        401: {"description": "No auth token provided"},
        403: {"description": "Admin privileges required"},
        404: {"description": "Submission not found"},
        422: {"description": "Submission is approved and cannot be deleted"},
    },
)
async def delete_submission(
    submission_id: str,
    current_user: dict = Depends(require_admin),
    db: Session = Depends(get_db),
) -> SubmissionDeleteResponse:
    """Permanently delete a pending or rejected submission."""
    service = SubmissionsService(db)
    return service.delete_submission(submission_id)

